lib/spack/spack/concretize.py

```diff
@@ -218,9 +218,11 @@ def concretize_architecture(self, spec):
         sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
         spec_changed = False
 
-        if spec.architecture is None:
+        if spec.architecture is None and root_arch is None:
```

Comment:
No, I have not checked the case. Good idea, I'll do something similar to what concretize_compiler does. I'll also work on adding more tests to this PR.